service/history/api/queryworkflow/api.go

```diff
@@ -345,13 +390,37 @@ func queryDirectlyThroughMatching(
 	nonStickyStartTime := time.Now().UTC()
 	matchingResp, err := matchingClient.QueryWorkflow(ctx, nonStickyMatchingRequest)
 	metrics.DirectQueryDispatchNonStickyLatency.With(metricsHandler).Record(time.Since(nonStickyStartTime))
-	if err != nil {
-		return nil, err
-	}
 	metrics.DirectQueryDispatchNonStickySuccessCount.With(metricsHandler).Record(1)
 	return &historyservice.QueryWorkflowResponse{
 		Response: &workflowservice.QueryWorkflowResponse{
 			QueryResult:   matchingResp.GetQueryResult(),
 			QueryRejected: matchingResp.GetQueryRejected(),
 		}}, err
 }
+
+func emitWorkflowQueryMetrics(
+	metricsHandler metrics.Handler,
+	nsEntry *namespace.Namespace,
+	msResp *historyservice.GetMutableStateResponse,
+	queryType string,
+	err error,
+) {
+	commonTags := []metrics.Tag{
+		metrics.OperationTag(metrics.HistoryQueryWorkflowScope),
+		metrics.NamespaceTag(nsEntry.Name().String()),
+		metrics.VersioningBehaviorTag(workflow.GetEffectiveVersioningBehavior(msResp.GetVersioningInfo())),
+		metrics.WorkflowStatusTag(msResp.GetWorkflowStatus().String()),
+		metrics.QueryTypeTag(queryType),
+	}
+
+	if err != nil {
+		if common.IsContextDeadlineExceededErr(err) {
+			metrics.WorkflowQueryTimeoutCount.With(metricsHandler).Record(1, commonTags...)
+		} else {
+			metrics.WorkflowQueryFailureCount.With(metricsHandler).Record(1, commonTags...)
+		}
+		return
+	}
+
+	metrics.WorkflowQuerySuccessCount.With(metricsHandler).Record(1, commonTags...)
```

Comment:
Nit: just slightly easier to read:
```suggestion
	if err == nil {
	        metrics.WorkflowQuerySuccessCount.With(metricsHandler).Record(1, commonTags...)
	} else  if common.IsContextDeadlineExceededErr(err) {
		metrics.WorkflowQueryTimeoutCount.With(metricsHandler).Record(1, commonTags...)
	} else {
		metrics.WorkflowQueryFailureCount.With(metricsHandler).Record(1, commonTags...)
	}
```